finality-aleph/src/validator_network/manager/mod.rs

```diff
@@ -30,191 +41,165 @@ impl Display for SendError {
+                false => missing_outgoing.insert(peer.clone()),
+            };
+        }
         ManagerStatus {
-            wanted_peers: manager.addresses.len(),
-            both_ways_peers: both_ways,
-            incoming_peers: incoming,
-            outgoing_peers: outgoing,
-            missing_peers: missing,
+            incoming_peers,
+            missing_incoming,
+            outgoing_peers,
+            missing_outgoing,
         }
     }
+
+    fn wanted_incoming(&self) -> usize {
+        self.incoming_peers.len() + self.missing_incoming.len()
+    }
+
+    fn wanted_outgoing(&self) -> usize {
+        self.outgoing_peers.len() + self.missing_outgoing.len()
+    }
+}
+
+fn pretty_authority_id_set(set: &HashSet<AuthorityId>) -> String {
+    set.iter()
+        .map(|authority_id| authority_id.to_short_string())
+        .collect::<Vec<_>>()
+        .join(", ")
 }
 
 impl Display for ManagerStatus {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
-        if self.wanted_peers == 0 {
+        let wanted_incoming = self.wanted_incoming();
+        let wanted_outgoing = self.wanted_outgoing();
+        if wanted_incoming + wanted_outgoing == 0 {
             return write!(f, "not maintaining any connections; ");
         }
 
-        write!(f, "target - {:?} connections; ", self.wanted_peers)?;
-
-        if self.both_ways_peers.is_empty() && self.incoming_peers.is_empty() {
-            write!(f, "WARNING! No incoming peers even though we expected tham, maybe connecting to us is impossible; ")?;
-        }
-
-        if !self.both_ways_peers.is_empty() {
-            let peers = self
-                .both_ways_peers
-                .iter()
-                .map(|authority_id| authority_id.to_short_string())
-                .collect::<Vec<_>>()
-                .join(", ");
-            write!(
-                f,
-                "both ways - {:?} [{}]; ",
-                self.both_ways_peers.len(),
-                peers,
-            )?;
-        }
-
-        if !self.incoming_peers.is_empty() {
-            let peers = self
-                .incoming_peers
-                .iter()
-                .map(|authority_id| authority_id.to_short_string())
-                .collect::<Vec<_>>()
-                .join(", ");
-            write!(
-                f,
-                "incoming only - {:?} [{}]; ",
-                self.incoming_peers.len(),
-                peers
-            )?;
-        }
-
-        if !self.outgoing_peers.is_empty() {
-            let peers = self
-                .outgoing_peers
-                .iter()
-                .map(|authority_id| authority_id.to_short_string())
-                .collect::<Vec<_>>()
-                .join(", ");
-            write!(
-                f,
-                "outgoing only - {:?} [{}];",
-                self.outgoing_peers.len(),
-                peers
-            )?;
+        match wanted_incoming {
```

Comment:
weak suggestion: merge both merges and the `if` above to `match (wanted_incoming, wanted_outgoing)`

Reply:
Hmm, maybe I don't understand the suggestion, but I don't think this simplifies anything – if I match on the pair I have to duplicate the nonempty printing code between `(0, _)` & `(_,0)` vs `(_,_)` cases, no?

I considered extracting separate methods for formatting incoming and outgoing outputs (essentially the two `match`es here), but I don't think this actually helps much either and, more importantly, seems relatively unrelated to your comment. ;p

Reply:
yeah, let's leave it as it is